controls/V-72315.rb

```diff
@@ -64,30 +120,69 @@
   forward-ports:
   icmp-blocks:
   rich rules:
- rule family=\"ipv4\" source address=\"92.188.21.1/24\" accept
- rule family=\"ipv4\" source address=\"211.17.142.46/32\" accept
-
+  rule family=\"ipv4\" source address=\"92.188.21.1/24\" accept
+  rule family=\"ipv4\" source address=\"211.17.142.46/32\" accept
 If \"firewalld\" is not active, determine whether \"tcpwrappers\" is being used by
 checking whether the \"hosts.allow\" and \"hosts.deny\" files are empty with the
 following commands:
-
 # ls -al /etc/hosts.allow
 rw-r----- 1 root root 9 Aug  2 23:13 /etc/hosts.allow
-
 # ls -al /etc/hosts.deny
 -rw-r----- 1 root root  9 Apr  9  2007 /etc/hosts.deny
-
 If \"firewalld\" and \"tcpwrappers\" are not installed, configured, and active, ask
 the SA if another access control program (such as iptables) is installed and active.
 Ask the SA to show that the running configuration grants or denies access to
 specific hosts or services.
-
 If \"firewalld\" is active and is not configured to grant access to specific hosts
 and \"tcpwrappers\" is not configured to grant or deny access to specific hosts,
 this is a finding."
   tag "fix": "If \"firewalld\" is installed and active on the system, configure
 rules for allowing specific services and hosts.
-
 If \"tcpwrappers\" is installed, configure the \"/etc/hosts.allow\" and
 \"/etc/hosts.deny\" to allow or deny access to specific hosts."
+
+  if service('firewalld').running?
+    @default_zone = firewalld.default_zone
+
+    describe firewalld.where{ zone = @default_zone } do
+      its('services') { should be_in FIREWALLD_SERVICES }
+    end
+
+    describe firewalld do
+      firewalld_hosts_allow.each do |rule|
+        it { should have_rule_enabled(rule) }
+      end
+      firewalld_hosts_deny.each do |rule|
+        it { should_not have_rule_enabled(rule) }
+      end
+      firewalld_ports_allow.each do |port|
+        it { should have_port_enabled_in_zone(port) }
+      end
+      firewalld_ports_deny.each do |port|
+        it { should_not have_port_enabled_in_zone(port) }
+      end
+    end
+  elsif service('iptables').running?
+    describe iptables do
+      iptable_rules.each do |rule|
+        it { should have_rule(rule) }
+      end
+    end
+  else
+    describe package('tcp_wrappers') do
+      it { should be_installed }
+    end
+    tcpwrappers_allow.each do |rule|
+      describe etc_hosts_allow.where { daemon == rule['daemon'] } do
+        its('client_list') { should include rule['client_list'] }
```

Comment:
This should probably be something like `be` instead of `include`. In theory, this should be a maxiumum set.

We're going to need a way to feed site specific data into these checks without modifying the files themselves because that's just not viable over time.

Reply:
@trevor-vaughan I changed these from "include" to "be". As far as site-specific data, can't you do that through the inputs that are defined in the control?